{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sequence Record\n",
    "\n",
    "Essentially a sequence with database metadata attached to it. Depending on the source, this metadata can be surprisingly rich."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: YP_025292.1\n",
      "Name: HokC\n",
      "Description: toxic membrane protein, small\n",
      "Number of features: 0\n",
      "Seq('MKQHKAMIVALIVICITAVVAALVTRKD', IUPACProtein())\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC\n",
    "record =SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKD\",\n",
    " IUPAC.protein),\n",
    " id=\"YP_025292.1\", name=\"HokC\",\n",
    " description=\"toxic membrane protein, small\")\n",
    "print record"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You get a `SeqRecord` when you:\n",
    "- parse a file with sequence data in one of the standard formats\n",
    "- obtain sequence data from an online resource (eg NCBI, UniProt, etc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Entrez Utilities - Sequence queries\n",
    "\n",
    "Online access to NCBI's data. The [Bio.Entrez](http://biopython.org/DIST/docs/api/Bio.Entrez-module.html) module makes use of the Entrez\n",
    "Programming Utilities (also known as EUtils), described in detail [here at NCBI](http://www.ncbi.nlm.nih.gov/entrez/utils/). You usually get an XML output, which is parsed using some utility function in `Bio.Entrez`.\n",
    "\n",
    "**Important:** Be aware that the Entrez utilities impose usage limits and you can get yourself blacklisted if you violate those. BioPython takes care of keeping an eye on those for you. You also should fill in a non-fake email address. See the [Frequency section](http://www.ncbi.nlm.nih.gov/books/NBK25497/#chapter2\\2e Frequency_Timing_and_Registrati) in the manual.\n",
    "\n",
    "# esearch\n",
    "\n",
    "`esearch` searches and retrieves primary IDs (for use in EFetch, ELink, and ESummary) and term translations and optionally retains results for future use in the user's environment. `efetch` can then be used to download the actual records.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Handle: <addinfourl at 31691048 whose fp = <socket._fileobject object at 0x1e35450>>\n",
      "\n",
      "\n",
      "Content: <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n",
      "<!DOCTYPE eSearchResult PUBLIC \"-//NLM//DTD esearch 20060628//EN\" \"http://eutils.ncbi.nlm.nih.gov/eutils/dtd/20060628/esearch.dtd\">\n",
      "<eSearchResult><Count>3</Count><RetMax>3</RetMax><RetStart>0</RetStart><IdList>\n",
      "<Id>377580661</Id>\n",
      "<Id>156535673</Id>\n",
      "<Id>156535671</Id>\n",
      "</IdList><TranslationSet><Translation>     <From>opuntia[ORGN]</From>     <To>\"Opuntia\"[Organism]</To>    </Translation></TranslationSet><TranslationStack>   <TermSet>    <Term>\"Opuntia\"[Organism]</Term>    <Field>Organism</Field>    <Count>1725</Count>    <Explode>Y</Explode>   </TermSet>   <TermSet>    <Term>accD[All Fields]</Term>    <Field>All Fields</Field>    <Count>161156</Count>    <Explode>N</Explode>   </TermSet>   <OP>AND</OP>  </TranslationStack><QueryTranslation>\"Opuntia\"[Organism] AND accD[All Fields]</QueryTranslation></eSearchResult>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = \"my_little_self@utu.fi\"\n",
    "handle = Entrez.esearch(db=\"nucleotide\", retmax=10, term=\"opuntia[ORGN] accD\")\n",
    "print \"Handle:\", handle #A file-like object, you can read it\n",
    "content=handle.read()\n",
    "print \"\\n\\nContent:\", content"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's XML allright. Let's parse it. The primary method to do this is `Entrez.read()` but it expects a file. We have already consumed our handle - we read the data from it to the end so that it cannot be used again. We could query again, or we can use `StringIO` which is a module which can simulate file objects on top of strings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsed result of equery: {u'Count': '3', u'RetMax': '3', u'IdList': ['377580661', '156535673', '156535671'], u'TranslationStack': [{u'Count': '1725', u'Field': 'Organism', u'Term': '\"Opuntia\"[Organism]', u'Explode': 'Y'}, {u'Count': '161156', u'Field': 'All Fields', u'Term': 'accD[All Fields]', u'Explode': 'N'}, 'AND'], u'TranslationSet': [{u'To': '\"Opuntia\"[Organism]', u'From': 'opuntia[ORGN]'}], u'RetStart': '0', u'QueryTranslation': '\"Opuntia\"[Organism] AND accD[All Fields]'}\n"
     ]
    }
   ],
   "source": [
    "import StringIO #Useful whenever you have your data in a string but a function which expects a file\n",
    "content_filelike=StringIO.StringIO(content)\n",
    "equery_result=Entrez.read(content_filelike)\n",
    "print \"Parsed result of equery:\", equery_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So we are having a dictionary. Most important here is the `IdList` key which holds the IDs which fit our query. Now we can fetch those.\n",
    "\n",
    "### read() vs. parse()\n",
    "\n",
    "Throughout BioPython in the different modules the function `read()` is used to parse an output which only has a single record (of whatever kind) and it returns that record plus it may even throw an error if the output seems to contain more than one record. On the other hand `parse()` is used to parse an output which has several records and returns a generator over them (ie something you can for-loop over once, or turn into a list with `list()`).\n",
    "\n",
    "## efetch\n",
    "\n",
    "Now that we have our IDs, we can fetch the records:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Efetch result <generator object parse at 0x1f32280>\n",
      "{u'GBSeq_moltype': 'DNA', u'GBSeq_source': 'chloroplast Opuntia decumbens', u'GBSeq_sequence': 'gagaccgggttacctgaagctattcaaacaggtataggtgaactaaacggtattcctttagcaattggtgttatggattttgggtttatagggggttctatgggagccgtagtaggtgaaaaaatcacccggttgattgaatatgctaccaaaaattcactaccccttattatagtatgtgcttctggaggggcacggatgcaagaaggcatcgtgagcttaatgcaaatggctaaaatatcttctgccttgcacgactatcactataaaccggtcgaaaaaaaattattatatatatcaattcttgcatcgcccactagtggtggtgtgacagccagttttggtatgttgggggatattattattgccgaaccagacgccgaaattgcgtttgcgggtaaaagagtaattgaggaagttttgaaggaagaagtacccgaaggtgcacaaacaaccgaaaatttattcgaaaatggtttattcgatccagtcttaccacgtaatcttttaaagagcgctctaagtgagttacttgagctgcacggtttttttcctttaaataaaacccaagccaag', u'GBSeq_primary-accession': 'HQ620723', u'GBSeq_definition': 'Opuntia decumbens voucher Martinez & Eggli 146a (ZSS) acetyl-CoA carboxylase beta subunit (accD) gene, partial cds; chloroplast', u'GBSeq_accession-version': 'HQ620723.1', u'GBSeq_topology': 'linear', u'GBSeq_length': '576', u'GBSeq_feature-table': [{u'GBFeature_quals': [{u'GBQualifier_name': 'organism', u'GBQualifier_value': 'Opuntia decumbens'}, {u'GBQualifier_name': 'organelle', u'GBQualifier_value': 'plastid:chloroplast'}, {u'GBQualifier_name': 'mol_type', u'GBQualifier_value': 'genomic DNA'}, {u'GBQualifier_name': 'specimen_voucher', u'GBQualifier_value': 'Martinez & Eggli 146a (ZSS)'}, {u'GBQualifier_name': 'db_xref', u'GBQualifier_value': 'taxon:867482'}, {u'GBQualifier_name': 'tissue_type', u'GBQualifier_value': 'stem'}, {u'GBQualifier_name': 'authority', u'GBQualifier_value': 'Opuntia decumbens Salm-Dyck'}], u'GBFeature_key': 'source', u'GBFeature_intervals': [{u'GBInterval_from': '1', u'GBInterval_to': '576', u'GBInterval_accession': 'HQ620723.1'}], u'GBFeature_location': '1..576'}, {u'GBFeature_quals': [{u'GBQualifier_name': 'gene', u'GBQualifier_value': 'accD'}], u'GBFeature_intervals': [{u'GBInterval_from': '1', u'GBInterval_to': '576', u'GBInterval_accession': 'HQ620723.1'}], u'GBFeature_location': '<1..>576', u'GBFeature_key': 'gene', u'GBFeature_partial5': StringElement('', attributes={u'value': u'true'}), u'GBFeature_partial3': StringElement('', attributes={u'value': u'true'})}, {u'GBFeature_quals': [{u'GBQualifier_name': 'gene', u'GBQualifier_value': 'accD'}, {u'GBQualifier_name': 'codon_start', u'GBQualifier_value': '1'}, {u'GBQualifier_name': 'transl_table', u'GBQualifier_value': '11'}, {u'GBQualifier_name': 'product', u'GBQualifier_value': 'acetyl-CoA carboxylase beta subunit'}, {u'GBQualifier_name': 'protein_id', u'GBQualifier_value': 'AFB70469.1'}, {u'GBQualifier_name': 'db_xref', u'GBQualifier_value': 'GI:377580662'}, {u'GBQualifier_name': 'translation', u'GBQualifier_value': 'ETGLPEAIQTGIGELNGIPLAIGVMDFGFIGGSMGAVVGEKITRLIEYATKNSLPLIIVCASGGARMQEGIVSLMQMAKISSALHDYHYKPVEKKLLYISILASPTSGGVTASFGMLGDIIIAEPDAEIAFAGKRVIEEVLKEEVPEGAQTTENLFENGLFDPVLPRNLLKSALSELLELHGFFPLNKTQAK'}], u'GBFeature_intervals': [{u'GBInterval_from': '1', u'GBInterval_to': '576', u'GBInterval_accession': 'HQ620723.1'}], u'GBFeature_location': '<1..>576', u'GBFeature_key': 'CDS', u'GBFeature_partial5': StringElement('', attributes={u'value': u'true'}), u'GBFeature_partial3': StringElement('', attributes={u'value': u'true'})}], u'GBSeq_create-date': '26-FEB-2012', u'GBSeq_other-seqids': ['gb|HQ620723.1|', 'gi|377580661'], u'GBSeq_division': 'PLN', u'GBSeq_taxonomy': 'Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta; Spermatophyta; Magnoliophyta; eudicotyledons; Gunneridae; Pentapetalae; Caryophyllales; Cactineae; Cactaceae; Opuntioideae; Opuntia', u'GBSeq_comment': '##Assembly-Data-START## Assembly Method MIRA V3rc4; Geneious v. 4.8 Sequencing Technology 454 ##Assembly-Data-END##', u'GBSeq_references': [{u'GBReference_position': '1..576', u'GBReference_reference': '1', u'GBReference_authors': ['Arakaki,M.', 'Christin,P.A.', 'Nyffeler,R.', 'Lendel,A.', 'Eggli,U.', 'Ogburn,R.M.', 'Spriggs,E.', 'Moore,M.J.', 'Edwards,E.J.'], u'GBReference_journal': 'Proc. Natl. Acad. Sci. U.S.A. 108 (20), 8379-8384 (2011)', u'GBReference_pubmed': '21536881', u'GBReference_xref': [{u'GBXref_id': '10.1073/pnas.1100628108', u'GBXref_dbname': 'doi'}], u'GBReference_title': \"Contemporaneous and recent radiations of the world's major succulent plant lineages\"}, {u'GBReference_position': '1..576', u'GBReference_reference': '2', u'GBReference_journal': 'Submitted (15-NOV-2010) Department of Ecology and Evolutionary Biology, Brown University, 80 Waterman St., Providence, RI 02912, USA', u'GBReference_title': 'Direct Submission', u'GBReference_authors': ['Arakaki,M.', 'Christin,P.-A.', 'Nyffeler,R.', 'Eggli,U.', 'Ogburn,R.M.', 'Spriggs,E.', 'Moore,M.J.', 'Edwards,E.J.']}], u'GBSeq_update-date': '26-FEB-2012', u'GBSeq_organism': 'Opuntia decumbens', u'GBSeq_locus': 'HQ620723', u'GBSeq_strandedness': 'double'}\n",
      "{u'GBSeq_moltype': 'DNA', u'GBSeq_source': 'chloroplast Opuntia humifusa', u'GBSeq_sequence': 'ggtgaaaaaatcacccggttgattgaatatgctaccaaaaattcactaccccttattatagtatgtgcttctggaggggcacggatgcaagaaggcattgtgagcttaatgcaaatggctaaaatatcttctgccttgtacgactatcactataaaccggtcgaaaaaaaattattatatatatcaattcttgcatcgcccactagtggtggtgtgacagccagttttggtatgttgggggatattattattgccgaaccagacgccgaaa', u'GBSeq_primary-accession': 'EF590893', u'GBSeq_definition': 'Opuntia humifusa voucher NMNH 06-8355 acetyl-CoA carboxylase-D (accD) gene, partial cds; chloroplast', u'GBSeq_accession-version': 'EF590893.1', u'GBSeq_topology': 'linear', u'GBSeq_length': '271', u'GBSeq_feature-table': [{u'GBFeature_quals': [{u'GBQualifier_name': 'organism', u'GBQualifier_value': 'Opuntia humifusa'}, {u'GBQualifier_name': 'organelle', u'GBQualifier_value': 'plastid:chloroplast'}, {u'GBQualifier_name': 'mol_type', u'GBQualifier_value': 'genomic DNA'}, {u'GBQualifier_name': 'specimen_voucher', u'GBQualifier_value': 'NMNH 06-8355'}, {u'GBQualifier_name': 'db_xref', u'GBQualifier_value': 'taxon:446152'}, {u'GBQualifier_name': 'collection_date', u'GBQualifier_value': 'Sep-2006'}, {u'GBQualifier_name': 'collected_by', u'GBQualifier_value': 'DL Erickson'}], u'GBFeature_key': 'source', u'GBFeature_intervals': [{u'GBInterval_from': '1', u'GBInterval_to': '271', u'GBInterval_accession': 'EF590893.1'}], u'GBFeature_location': '1..271'}, {u'GBFeature_quals': [{u'GBQualifier_name': 'gene', u'GBQualifier_value': 'accD'}], u'GBFeature_intervals': [{u'GBInterval_from': '1', u'GBInterval_to': '271', u'GBInterval_accession': 'EF590893.1'}], u'GBFeature_location': '<1..>271', u'GBFeature_key': 'gene', u'GBFeature_partial5': StringElement('', attributes={u'value': u'true'}), u'GBFeature_partial3': StringElement('', attributes={u'value': u'true'})}, {u'GBFeature_quals': [{u'GBQualifier_name': 'gene', u'GBQualifier_value': 'accD'}, {u'GBQualifier_name': 'codon_start', u'GBQualifier_value': '1'}, {u'GBQualifier_name': 'transl_table', u'GBQualifier_value': '11'}, {u'GBQualifier_name': 'product', u'GBQualifier_value': 'acetyl-CoA carboxylase-D'}, {u'GBQualifier_name': 'protein_id', u'GBQualifier_value': 'ABU79977.1'}, {u'GBQualifier_name': 'db_xref', u'GBQualifier_value': 'GI:156535674'}, {u'GBQualifier_name': 'translation', u'GBQualifier_value': 'GEKITRLIEYATKNSLPLIIVCASGGARMQEGIVSLMQMAKISSALYDYHYKPVEKKLLYISILASPTSGGVTASFGMLGDIIIAEPDAE'}], u'GBFeature_intervals': [{u'GBInterval_from': '1', u'GBInterval_to': '271', u'GBInterval_accession': 'EF590893.1'}], u'GBFeature_location': '<1..>271', u'GBFeature_key': 'CDS', u'GBFeature_partial5': StringElement('', attributes={u'value': u'true'}), u'GBFeature_partial3': StringElement('', attributes={u'value': u'true'})}], u'GBSeq_create-date': '02-SEP-2007', u'GBSeq_other-seqids': ['gb|EF590893.1|', 'gi|156535673'], u'GBSeq_division': 'PLN', u'GBSeq_taxonomy': 'Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta; Spermatophyta; Magnoliophyta; eudicotyledons; Gunneridae; Pentapetalae; Caryophyllales; Cactineae; Cactaceae; Opuntioideae; Opuntia', u'GBSeq_references': [{u'GBReference_position': '1..271', u'GBReference_reference': '1', u'GBReference_authors': ['Kress,W.J.', 'Erickson,D.L.'], u'GBReference_journal': 'PLoS ONE 2 (6), E508 (2007)', u'GBReference_pubmed': '17551588', u'GBReference_xref': [{u'GBXref_id': '10.1371/journal.pone.0000508', u'GBXref_dbname': 'doi'}], u'GBReference_title': 'A two-locus global DNA barcode for land plants: the coding rbcL gene complements the non-coding trnH-psbA spacer region', u'GBReference_remark': 'Publication Status: Online-Only'}, {u'GBReference_position': '1..271', u'GBReference_reference': '2', u'GBReference_journal': 'Submitted (03-MAY-2007) Botany, Smithsonian Institution, National Museum of Natural History, Washington, DC 20013-7012, USA', u'GBReference_title': 'Direct Submission', u'GBReference_authors': ['Kress,J.', 'Erickson,D.']}], u'GBSeq_update-date': '02-SEP-2007', u'GBSeq_organism': 'Opuntia humifusa', u'GBSeq_locus': 'EF590893', u'GBSeq_strandedness': 'double'}\n",
      "{u'GBSeq_moltype': 'DNA', u'GBSeq_source': 'chloroplast Opuntia fragilis', u'GBSeq_sequence': 'ggtgaaaaaatcacccggttgattgaatatgctaccaaaaattcactaccccttattatagtatgtgcttctggaggggcacggatgcaagaaggcatcgtgagcttaatgcaaatggctaaaatatcttctgccttgtacgactatcaccataaaccggttgaaaaaaaattgttatatatatcaattcttgcatcgcccactagtggtggtgtgacagccagttttggtatgttgggggatattattattgccgaaccagacgccgaaa', u'GBSeq_primary-accession': 'EF590892', u'GBSeq_definition': 'Opuntia fragilis voucher USBG s.n. acetyl-CoA carboxylase-D (accD) gene, partial cds; chloroplast', u'GBSeq_accession-version': 'EF590892.1', u'GBSeq_topology': 'linear', u'GBSeq_length': '271', u'GBSeq_feature-table': [{u'GBFeature_quals': [{u'GBQualifier_name': 'organism', u'GBQualifier_value': 'Opuntia fragilis'}, {u'GBQualifier_name': 'organelle', u'GBQualifier_value': 'plastid:chloroplast'}, {u'GBQualifier_name': 'mol_type', u'GBQualifier_value': 'genomic DNA'}, {u'GBQualifier_name': 'specimen_voucher', u'GBQualifier_value': 'USBG s.n.'}, {u'GBQualifier_name': 'db_xref', u'GBQualifier_value': 'taxon:446151'}, {u'GBQualifier_name': 'collection_date', u'GBQualifier_value': 'Sep-2006'}, {u'GBQualifier_name': 'collected_by', u'GBQualifier_value': 'DL Erickson'}], u'GBFeature_key': 'source', u'GBFeature_intervals': [{u'GBInterval_from': '1', u'GBInterval_to': '271', u'GBInterval_accession': 'EF590892.1'}], u'GBFeature_location': '1..271'}, {u'GBFeature_quals': [{u'GBQualifier_name': 'gene', u'GBQualifier_value': 'accD'}], u'GBFeature_intervals': [{u'GBInterval_from': '1', u'GBInterval_to': '271', u'GBInterval_accession': 'EF590892.1'}], u'GBFeature_location': '<1..>271', u'GBFeature_key': 'gene', u'GBFeature_partial5': StringElement('', attributes={u'value': u'true'}), u'GBFeature_partial3': StringElement('', attributes={u'value': u'true'})}, {u'GBFeature_quals': [{u'GBQualifier_name': 'gene', u'GBQualifier_value': 'accD'}, {u'GBQualifier_name': 'codon_start', u'GBQualifier_value': '1'}, {u'GBQualifier_name': 'transl_table', u'GBQualifier_value': '11'}, {u'GBQualifier_name': 'product', u'GBQualifier_value': 'acetyl-CoA carboxylase-D'}, {u'GBQualifier_name': 'protein_id', u'GBQualifier_value': 'ABU79976.1'}, {u'GBQualifier_name': 'db_xref', u'GBQualifier_value': 'GI:156535672'}, {u'GBQualifier_name': 'translation', u'GBQualifier_value': 'GEKITRLIEYATKNSLPLIIVCASGGARMQEGIVSLMQMAKISSALYDYHHKPVEKKLLYISILASPTSGGVTASFGMLGDIIIAEPDAE'}], u'GBFeature_intervals': [{u'GBInterval_from': '1', u'GBInterval_to': '271', u'GBInterval_accession': 'EF590892.1'}], u'GBFeature_location': '<1..>271', u'GBFeature_key': 'CDS', u'GBFeature_partial5': StringElement('', attributes={u'value': u'true'}), u'GBFeature_partial3': StringElement('', attributes={u'value': u'true'})}], u'GBSeq_create-date': '02-SEP-2007', u'GBSeq_other-seqids': ['gb|EF590892.1|', 'gi|156535671'], u'GBSeq_division': 'PLN', u'GBSeq_taxonomy': 'Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta; Spermatophyta; Magnoliophyta; eudicotyledons; Gunneridae; Pentapetalae; Caryophyllales; Cactineae; Cactaceae; Opuntioideae; Opuntia', u'GBSeq_references': [{u'GBReference_position': '1..271', u'GBReference_reference': '1', u'GBReference_authors': ['Kress,W.J.', 'Erickson,D.L.'], u'GBReference_journal': 'PLoS ONE 2 (6), E508 (2007)', u'GBReference_pubmed': '17551588', u'GBReference_xref': [{u'GBXref_id': '10.1371/journal.pone.0000508', u'GBXref_dbname': 'doi'}], u'GBReference_title': 'A two-locus global DNA barcode for land plants: the coding rbcL gene complements the non-coding trnH-psbA spacer region', u'GBReference_remark': 'Publication Status: Online-Only'}, {u'GBReference_position': '1..271', u'GBReference_reference': '2', u'GBReference_journal': 'Submitted (03-MAY-2007) Botany, Smithsonian Institution, National Museum of Natural History, Washington, DC 20013-7012, USA', u'GBReference_title': 'Direct Submission', u'GBReference_authors': ['Kress,J.', 'Erickson,D.']}], u'GBSeq_update-date': '02-SEP-2007', u'GBSeq_organism': 'Opuntia fragilis', u'GBSeq_locus': 'EF590892', u'GBSeq_strandedness': 'double'}\n"
     ]
    }
   ],
   "source": [
    "handle=Entrez.efetch(\"nucleotide\",id=equery_result['IdList'],retmode=\"xml\")\n",
    "efetch_result=Entrez.parse(handle)\n",
    "print \"Efetch result\", efetch_result #This gives a \"generator\". You may want to make this into a list, or loop over it\n",
    "for r in efetch_result:\n",
    "    print r"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is **not** what we wanted! What we want is a `SeqRecord` object. Reason: `Entrez.parse()` is a generic function for parsing Entrez XML. We are receiving nucleotide database records, ie sequences. For this, we want [Bio.SeqIO](http://biopython.org/DIST/docs/_api_159/Bio.SeqIO-module.html) which is specialized in parsing the various sequence formats we can meet and in this particular case the *GenBank* format (`gb`  is what we want)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Efetch result <generator object parse at 0x3299f00>\n",
      "Opuntia decumbens voucher Martinez & Eggli 146a (ZSS) acetyl-CoA carboxylase beta subunit (accD) gene, partial cds; chloroplast.\n",
      "SeqRecord(seq=Seq('GAGACCGGGTTACCTGAAGCTATTCAAACAGGTATAGGTGAACTAAACGGTATT...AAG', IUPACAmbiguousDNA()), id='HQ620723.1', name='HQ620723', description='Opuntia decumbens voucher Martinez & Eggli 146a (ZSS) acetyl-CoA carboxylase beta subunit (accD) gene, partial cds; chloroplast.', dbxrefs=[])\n",
      "\n",
      "\n",
      "Opuntia humifusa voucher NMNH 06-8355 acetyl-CoA carboxylase-D (accD) gene, partial cds; chloroplast.\n",
      "SeqRecord(seq=Seq('GGTGAAAAAATCACCCGGTTGATTGAATATGCTACCAAAAATTCACTACCCCTT...AAA', IUPACAmbiguousDNA()), id='EF590893.1', name='EF590893', description='Opuntia humifusa voucher NMNH 06-8355 acetyl-CoA carboxylase-D (accD) gene, partial cds; chloroplast.', dbxrefs=[])\n",
      "\n",
      "\n",
      "Opuntia fragilis voucher USBG s.n. acetyl-CoA carboxylase-D (accD) gene, partial cds; chloroplast.\n",
      "SeqRecord(seq=Seq('GGTGAAAAAATCACCCGGTTGATTGAATATGCTACCAAAAATTCACTACCCCTT...AAA', IUPACAmbiguousDNA()), id='EF590892.1', name='EF590892', description='Opuntia fragilis voucher USBG s.n. acetyl-CoA carboxylase-D (accD) gene, partial cds; chloroplast.', dbxrefs=[])\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import Bio.SeqIO as SeqIO\n",
    "handle=Entrez.efetch(\"nucleotide\",id=equery_result['IdList'],rettype=\"gb\") #Note: this now says \"gb\" to get the right format\n",
    "efetch_result=SeqIO.parse(handle,\"gb\") #Note: this used to read Entrez.parse(handle)\n",
    "print \"Efetch result\", efetch_result #This gives a \"generator\". You may want to make this into a list, or loop over it\n",
    "for r in efetch_result:\n",
    "    print r.description\n",
    "    print repr(r)\n",
    "    print\n",
    "    print"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Much better. Now we have SeqRecords. Let us summarize everything into a nice piece of code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SeqRecord(seq=Seq('GAGACCGGGTTACCTGAAGCTATTCAAACAGGTATAGGTGAACTAAACGGTATT...AAG', IUPACAmbiguousDNA()), id='HQ620723.1', name='HQ620723', description='Opuntia decumbens voucher Martinez & Eggli 146a (ZSS) acetyl-CoA carboxylase beta subunit (accD) gene, partial cds; chloroplast.', dbxrefs=[])\n",
      "SeqRecord(seq=Seq('GGTGAAAAAATCACCCGGTTGATTGAATATGCTACCAAAAATTCACTACCCCTT...AAA', IUPACAmbiguousDNA()), id='EF590893.1', name='EF590893', description='Opuntia humifusa voucher NMNH 06-8355 acetyl-CoA carboxylase-D (accD) gene, partial cds; chloroplast.', dbxrefs=[])\n",
      "SeqRecord(seq=Seq('GGTGAAAAAATCACCCGGTTGATTGAATATGCTACCAAAAATTCACTACCCCTT...AAA', IUPACAmbiguousDNA()), id='EF590892.1', name='EF590892', description='Opuntia fragilis voucher USBG s.n. acetyl-CoA carboxylase-D (accD) gene, partial cds; chloroplast.', dbxrefs=[])\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "Entrez.email = \"my_little_self@utu.fi\"\n",
    "\n",
    "def fetch_records(query):\n",
    "    search_result_handle=Entrez.esearch(db=\"nucleotide\", retmax=10, term=query) #Query Entrez\n",
    "    search_result=Entrez.read(search_result_handle) #Parse the XML data you get in reply\n",
    "    efetch_result_handle=Entrez.efetch(db=\"nucleotide\", id=search_result['IdList'],rettype=\"gb\") #Now fetch the actual sequences in GenBank format\n",
    "    seq_records=SeqIO.parse(efetch_result_handle,format=\"gb\") #Now parse the obtained data using SeqIO\n",
    "    return seq_records\n",
    "\n",
    "for seq_rec in fetch_records(\"opuntia[ORGN] accD\"):\n",
    "    print repr(seq_rec)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Search & download recap (1)\n",
    "\n",
    "- `Entrez.esearch()` to get a reply with list of IDs\n",
    "- `Entrez.efetch()` to grab the actual records\n",
    "- `SeqIO.parse()` to obtain a sequence of SeqRecord objects\n",
    "\n",
    "This is okay for simple small queries but is highly discouraged for large queries. NCBI expects you to use the *query history* functionality which lets them bind your fetch to a previous search and use their internal caching mechanisms. Turn it on with `usehistory=\"y\"` like such:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Search_result: {u'Count': '3', u'RetMax': '3', u'IdList': ['377580661', '156535673', '156535671'], u'TranslationStack': [{u'Count': '1725', u'Field': 'Organism', u'Term': '\"Opuntia\"[Organism]', u'Explode': 'Y'}, {u'Count': '161156', u'Field': 'All Fields', u'Term': 'accD[All Fields]', u'Explode': 'N'}, 'AND'], u'QueryTranslation': '\"Opuntia\"[Organism] AND accD[All Fields]', u'TranslationSet': [{u'To': '\"Opuntia\"[Organism]', u'From': 'opuntia[ORGN]'}], u'RetStart': '0', u'QueryKey': '1', u'WebEnv': 'NCID_1_949678182_130.14.18.34_9001_1452672377_457968480_0MetA0_S_MegaStore_F_1'}\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "Entrez.email = \"my_little_self@utu.fi\"\n",
    "\n",
    "search_result_handle=Entrez.esearch(db=\"nucleotide\", retmax=10, term=\"opuntia[ORGN] accD\", usehistory=\"y\") #Query Entrez with history on\n",
    "search_result=Entrez.read(search_result_handle) #Parse the XML data you get in reply\n",
    "print \"Search_result:\", search_result\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note the `QueryKey` and `WebEnv` keys. This is the information which identifies your result set at NCBI's server and can be used to fetch the records like so:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SeqRecord(seq=Seq('GAGACCGGGTTACCTGAAGCTATTCAAACAGGTATAGGTGAACTAAACGGTATT...AAG', IUPACAmbiguousDNA()), id='HQ620723.1', name='HQ620723', description='Opuntia decumbens voucher Martinez & Eggli 146a (ZSS) acetyl-CoA carboxylase beta subunit (accD) gene, partial cds; chloroplast.', dbxrefs=[])\n",
      "\n",
      "\n",
      "SeqRecord(seq=Seq('GGTGAAAAAATCACCCGGTTGATTGAATATGCTACCAAAAATTCACTACCCCTT...AAA', IUPACAmbiguousDNA()), id='EF590893.1', name='EF590893', description='Opuntia humifusa voucher NMNH 06-8355 acetyl-CoA carboxylase-D (accD) gene, partial cds; chloroplast.', dbxrefs=[])\n",
      "\n",
      "\n",
      "SeqRecord(seq=Seq('GGTGAAAAAATCACCCGGTTGATTGAATATGCTACCAAAAATTCACTACCCCTT...AAA', IUPACAmbiguousDNA()), id='EF590892.1', name='EF590892', description='Opuntia fragilis voucher USBG s.n. acetyl-CoA carboxylase-D (accD) gene, partial cds; chloroplast.', dbxrefs=[])\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "efetch_result_handle=Entrez.efetch(db=\"nucleotide\",\n",
    "                                   query_key=search_result[\"QueryKey\"],\n",
    "                                   webenv=search_result[\"WebEnv\"],\n",
    "                                   rettype=\"gb\") #Now fetch the actual sequences in GenBank format. Note how I don't give the IDs!\n",
    "seq_records=SeqIO.parse(efetch_result_handle,format=\"gb\") #Now parse the obtained data using SeqIO\n",
    "for r in seq_records:\n",
    "    print repr(r)\n",
    "    print\n",
    "    print\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Search & download recap (2)\n",
    "\n",
    "- `Entrez.esearch(..., usehistory=\"y\")` to get a reply and webenv + query_key\n",
    "- `Entrez.efetch(..., webenv=, query_key=)` to grab the actual records from the previous search\n",
    "- `SeqIO.parse()` to obtain a sequence of SeqRecord objects\n",
    "\n",
    "Great! Now our problem is that if we have a large result set, thousands of IDs, Entrez doesn't let us grab them all at once with a single efetch. We need to get them bit-by-bit. Here's a complete example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetched 50 records\n",
      "First few: [SeqRecord(seq=UnknownSeq(3422749, alphabet = IUPACAmbiguousDNA(), character = 'N'), id='NZ_KQ130471.1', name='NZ_KQ130471', description='Sphingobium baderi LL03 genomic scaffold 25433_scaffold_01, whole genome shotgun sequence.', dbxrefs=['BioProject:PRJNA224116 BioSample:SAMN02899865 Assembly:GCF_001046655.1']), SeqRecord(seq=UnknownSeq(87969, alphabet = IUPACAmbiguousDNA(), character = 'N'), id='NZ_LMWF01000229.1', name='NZ_LMWF01000229', description='Streptomyces rimosus subsp. rimosus strain NRRL WC-3897 P254contig30.1, whole genome shotgun sequence.', dbxrefs=['BioProject:PRJNA224116 BioSample:SAMN02900606 Assembly:GCF_001507395.1']), SeqRecord(seq=Seq('GTCATATCCCAGTGCTGACTCCCGGGCGTGCAGACCGCTAACTAGCTCACTCGC...AAA', IUPACAmbiguousDNA()), id='NM_001318470.1', name='NM_001318470', description='Homo sapiens TSC22 domain family member 3 (TSC22D3), transcript variant 5, mRNA.', dbxrefs=[]), SeqRecord(seq=Seq('GTCATATCCCAGTGCTGACTCCCGGGCGTGCAGACCGCTAACTAGCTCACTCGC...AAA', IUPACAmbiguousDNA()), id='NM_001318468.1', name='NM_001318468', description='Homo sapiens TSC22 domain family member 3 (TSC22D3), transcript variant 4, mRNA.', dbxrefs=[]), SeqRecord(seq=UnknownSeq(5132365, alphabet = IUPACAmbiguousDNA(), character = 'N'), id='NZ_LN600312.1', name='NZ_LN600312', description='Xanthomonas citri pv. citri genome assembly ASX_S_PRJEB7150_v1, scaffold ASX_S_XAC908_scaffold1, whole genome shotgun sequence.', dbxrefs=['BioProject:PRJNA224116 BioSample:SAMEA2752228 Assembly:GCF_001498875.1'])]\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "Entrez.email = \"my_little_self@utu.fi\"\n",
    "\n",
    "def fetch_records(query, max_results=10, fetch_batch=5):\n",
    "    search_result_handle=Entrez.esearch(db=\"nucleotide\", retmax=max_results, term=query, usehistory=\"y\") #Query Entrez\n",
    "    search_result=Entrez.read(search_result_handle) #Parse the XML data I get in reply\n",
    "    #The only thing you care about is Count, and the Webenv+QueryKey data\n",
    "    all_records=[] #This will be our result list\n",
    "    #Divide into blocks of fetch_batch results, download one at a time\n",
    "    for start_index in range(0,len(search_result[\"IdList\"]),fetch_batch):\n",
    "        efetch_result_handle=Entrez.efetch(db=\"nucleotide\", \n",
    "                                           retstart=start_index,\n",
    "                                           retmax=fetch_batch,\n",
    "                                           webenv=search_result[\"WebEnv\"],\n",
    "                                           query_key=search_result[\"QueryKey\"],\n",
    "                                           rettype=\"gb\") #Now fetch the actual sequences in GenBank format\n",
    "        #add these records into our list\n",
    "        all_records.extend(SeqIO.parse(efetch_result_handle,format=\"gb\")) #Now parse the obtained data using SeqIO\n",
    "    return all_records\n",
    "\n",
    "fetched_records_p53=fetch_records(\"p53\", max_results=50, fetch_batch=10) #get max 50 records, fetch 10 at a time\n",
    "print \"Fetched\", len(fetched_records_p53), \"records\"\n",
    "print \"First few:\", fetched_records_p53[:5]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Search & download - final version\n",
    "\n",
    "- `Entrez.esearch(..., retmax=, usehistory=\"y\")` to get a reply and webenv + query_key, searches for retmax records\n",
    "- `for batch_idx in range(0, total_num_of_results, download_batch)` to chop the results into download_batch sized chunks\n",
    "- `Entrez.efetch(retstart=, retmax=download_batch, webenv=, query_key=)` to grab one batch at a time\n",
    "- `SeqIO.parse()` to obtain a sequence of SeqRecord objects for every batch"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's grab a protein and see what the record has:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SeqRecord(seq=Seq('MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWF...DSD', IUPACProtein()), id='NP_000537.3', name='NP_000537', description='cellular tumor antigen p53 isoform a [Homo sapiens].', dbxrefs=[])\n",
      "MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD\n"
     ]
    }
   ],
   "source": [
    "tp53_rec=SeqIO.read(Entrez.efetch(db=\"protein\",id=[\"NP_000537\"],rettype=\"gb\"),\"gb\")\n",
    "print repr(tp53_rec)\n",
    "print str(tp53_rec.seq)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- `seq` sequence itself, Seq object\n",
    "- `id` ID used to identify the sequence, accession number, string\n",
    "- `name` name for the sequence, string\n",
    "- `description` readable description or expressive name for the sequence, string\n",
    "- `letter_annotations` dictionary of additional information about the letters in the sequence\n",
    "- `annotations` dictionary of additional information about the sequence\n",
    "- `features` A list of [SeqFeature](http://biopython.org/DIST/docs/api/Bio.SeqFeature.SeqFeature-class.html) objects with information about the features on a sequence\n",
    "- `dbxrefs` A list of database cross-references as strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on SeqRecord in module Bio.SeqRecord object:\n",
      "\n",
      "class SeqRecord(__builtin__.object)\n",
      " |  A SeqRecord object holds a sequence and information about it.\n",
      " |  \n",
      " |  Main attributes:\n",
      " |      - id          - Identifier such as a locus tag (string)\n",
      " |      - seq         - The sequence itself (Seq object or similar)\n",
      " |  \n",
      " |  Additional attributes:\n",
      " |      - name        - Sequence name, e.g. gene name (string)\n",
      " |      - description - Additional text (string)\n",
      " |      - dbxrefs     - List of database cross references (list of strings)\n",
      " |      - features    - Any (sub)features defined (list of SeqFeature objects)\n",
      " |      - annotations - Further information about the whole sequence (dictionary).\n",
      " |        Most entries are strings, or lists of strings.\n",
      " |      - letter_annotations - Per letter/symbol annotation (restricted\n",
      " |        dictionary). This holds Python sequences (lists, strings\n",
      " |        or tuples) whose length matches that of the sequence.\n",
      " |        A typical use would be to hold a list of integers\n",
      " |        representing sequencing quality scores, or a string\n",
      " |        representing the secondary structure.\n",
      " |  \n",
      " |  You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |  SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |  objects directly (see the __init__ method for further details):\n",
      " |  \n",
      " |  >>> from Bio.Seq import Seq\n",
      " |  >>> from Bio.SeqRecord import SeqRecord\n",
      " |  >>> from Bio.Alphabet import IUPAC\n",
      " |  >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      " |  ...                         IUPAC.protein),\n",
      " |  ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |  ...                    description=\"toxic membrane protein\")\n",
      " |  >>> print(record)\n",
      " |  ID: YP_025292.1\n",
      " |  Name: HokC\n",
      " |  Description: toxic membrane protein\n",
      " |  Number of features: 0\n",
      " |  Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      " |  \n",
      " |  If you want to save SeqRecord objects to a sequence file, use Bio.SeqIO\n",
      " |  for this.  For the special case where you want the SeqRecord turned into\n",
      " |  a string in a particular file format there is a format method which uses\n",
      " |  Bio.SeqIO internally:\n",
      " |  \n",
      " |  >>> print(record.format(\"fasta\"))\n",
      " |  >YP_025292.1 toxic membrane protein\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  <BLANKLINE>\n",
      " |  \n",
      " |  You can also do things like slicing a SeqRecord, checking its length, etc\n",
      " |  \n",
      " |  >>> len(record)\n",
      " |  44\n",
      " |  >>> edited = record[:10] + record[11:]\n",
      " |  >>> print(edited.seq)\n",
      " |  MKQHKAMIVAIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  >>> print(record.seq)\n",
      " |  MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Add another sequence or string to this sequence.\n",
      " |      \n",
      " |      The other sequence can be a SeqRecord object, a Seq object (or\n",
      " |      similar, e.g. a MutableSeq) or a plain Python string. If you add\n",
      " |      a plain string or a Seq (like) object, the new SeqRecord will simply\n",
      " |      have this appended to the existing data. However, any per letter\n",
      " |      annotation will be lost:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = record + \"ACT\"\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNNACT\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |      \n",
      " |      The new record will attempt to combine the annotation, but for any\n",
      " |      ambiguities (e.g. different names) it defaults to omitting that\n",
      " |      annotation.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      \n",
      " |      Now let's cut the plasmid into two pieces, and join them back up the\n",
      " |      other way round (i.e. shift the starting point on this plasmid, have\n",
      " |      a look at the annotated features in the original file to see why this\n",
      " |      particular split point might make sense):\n",
      " |      \n",
      " |      >>> left = plasmid[:3765]\n",
      " |      >>> right = plasmid[3765:]\n",
      " |      >>> new = right + left\n",
      " |      >>> print(\"%s %i\" % (new.id, len(new)))\n",
      " |      pBAD30 4923\n",
      " |      >>> str(new.seq) == str(right.seq + left.seq)\n",
      " |      True\n",
      " |      >>> len(new.features) == len(left.features) + len(right.features)\n",
      " |      True\n",
      " |      \n",
      " |      When we add the left and right SeqRecord objects, their annotation\n",
      " |      is all consistent, so it is all conserved in the new SeqRecord:\n",
      " |      \n",
      " |      >>> new.id == left.id == right.id == plasmid.id\n",
      " |      True\n",
      " |      >>> new.name == left.name == right.name == plasmid.name\n",
      " |      True\n",
      " |      >>> new.description == plasmid.description\n",
      " |      True\n",
      " |      >>> new.annotations == left.annotations == right.annotations\n",
      " |      True\n",
      " |      >>> new.letter_annotations == plasmid.letter_annotations\n",
      " |      True\n",
      " |      >>> new.dbxrefs == left.dbxrefs == right.dbxrefs\n",
      " |      True\n",
      " |      \n",
      " |      However, we should point out that when we sliced the SeqRecord,\n",
      " |      any annotations dictionary or dbxrefs list entries were lost.\n",
      " |      You can explicitly copy them like this:\n",
      " |      \n",
      " |      >>> new.annotations = plasmid.annotations.copy()\n",
      " |      >>> new.dbxrefs = plasmid.dbxrefs[:]\n",
      " |  \n",
      " |  __bool__(self)\n",
      " |      Boolean value of an instance of this class (True).\n",
      " |      \n",
      " |      This behaviour is for backwards compatibility, since until the\n",
      " |      __len__ method was added, a SeqRecord always evaluated as True.\n",
      " |      \n",
      " |      Note that in comparison, a Seq object will evaluate to False if it\n",
      " |      has a zero length sequence.\n",
      " |      \n",
      " |      WARNING: The SeqRecord may in future evaluate to False when its\n",
      " |      sequence is of zero length (in order to better match the Seq\n",
      " |      object behaviour)!\n",
      " |  \n",
      " |  __contains__(self, char)\n",
      " |      Implements the 'in' keyword, searches the sequence.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> \"GAATTC\" in record\n",
      " |      False\n",
      " |      >>> \"AAA\" in record\n",
      " |      True\n",
      " |      \n",
      " |      This essentially acts as a proxy for using \"in\" on the sequence:\n",
      " |      \n",
      " |      >>> \"GAATTC\" in record.seq\n",
      " |      False\n",
      " |      >>> \"AAA\" in record.seq\n",
      " |      True\n",
      " |      \n",
      " |      Note that you can also use Seq objects as the query,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> Seq(\"AAA\") in record\n",
      " |      True\n",
      " |      >>> Seq(\"AAA\", generic_dna) in record\n",
      " |      True\n",
      " |      \n",
      " |      See also the Seq object's __contains__ method.\n",
      " |  \n",
      " |  __format__(self, format_spec)\n",
      " |      Returns the record as a string in the specified file format.\n",
      " |      \n",
      " |      This method supports the python format() function added in\n",
      " |      Python 2.6/3.0.  The format_spec should be a lower case string\n",
      " |      supported by Bio.SeqIO as an output file format. See also the\n",
      " |      SeqRecord's format() method.\n",
      " |      \n",
      " |      Under Python 3 please note that for binary formats a bytes\n",
      " |      string is returned, otherwise a (unicode) string is returned.\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Returns a sub-sequence or an individual letter.\n",
      " |      \n",
      " |      Slicing, e.g. my_record[5:10], returns a new SeqRecord for\n",
      " |      that sub-sequence with appropriate annotation preserved.  The\n",
      " |      name, id and description are kept.\n",
      " |      \n",
      " |      Any per-letter-annotations are sliced to match the requested\n",
      " |      sub-sequence.  Unless a stride is used, all those features\n",
      " |      which fall fully within the subsequence are included (with\n",
      " |      their locations adjusted accordingly).\n",
      " |      \n",
      " |      However, the annotations dictionary and the dbxrefs list are\n",
      " |      not used for the new SeqRecord, as in general they may not\n",
      " |      apply to the subsequence.  If you want to preserve them, you\n",
      " |      must explicitly copy them to the new SeqRecord yourself.\n",
      " |      \n",
      " |      Using an integer index, e.g. my_record[5] is shorthand for\n",
      " |      extracting that letter from the sequence, my_record.seq[5].\n",
      " |      \n",
      " |      For example, consider this short protein and its secondary\n",
      " |      structure as encoded by the PDB (e.g. H for alpha helices),\n",
      " |      plus a simple feature for its histidine self phosphorylation\n",
      " |      site:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> rec = SeqRecord(Seq(\"MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLAT\"\n",
      " |      ...                     \"EMMSEQDGYLAESINKDIEECNAIIEQFIDYLR\",\n",
      " |      ...                     IUPAC.protein),\n",
      " |      ...                 id=\"1JOY\", name=\"EnvZ\",\n",
      " |      ...                 description=\"Homodimeric domain of EnvZ from E. coli\")\n",
      " |      >>> rec.letter_annotations[\"secondary_structure\"] = \"  S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  \"\n",
      " |      >>> rec.features.append(SeqFeature(FeatureLocation(20, 21),\n",
      " |      ...                     type = \"Site\"))\n",
      " |      \n",
      " |      Now let's have a quick look at the full record,\n",
      " |      \n",
      " |      >>> print(rec)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR', IUPACProtein())\n",
      " |      >>> print(rec.letter_annotations[\"secondary_structure\"])\n",
      " |        S  SSSSSSHHHHHTTTHHHHHHHHHHHHHHHHHHHHHHTHHHHHHHHHHHHHHHHHHHHHTT  \n",
      " |      >>> print(rec.features[0].location)\n",
      " |      [20:21]\n",
      " |      \n",
      " |      Now let's take a sub sequence, here chosen as the first (fractured)\n",
      " |      alpha helix which includes the histidine phosphorylation site:\n",
      " |      \n",
      " |      >>> sub = rec[11:41]\n",
      " |      >>> print(sub)\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('RTLLMAGVSHDLRTPLTRIRLATEMMSEQD', IUPACProtein())\n",
      " |      >>> print(sub.letter_annotations[\"secondary_structure\"])\n",
      " |      HHHHHTTTHHHHHHHHHHHHHHHHHHHHHH\n",
      " |      >>> print(sub.features[0].location)\n",
      " |      [9:10]\n",
      " |      \n",
      " |      You can also of course omit the start or end values, for\n",
      " |      example to get the first ten letters only:\n",
      " |      \n",
      " |      >>> print(rec[:10])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLAD', IUPACProtein())\n",
      " |      \n",
      " |      Or for the last ten letters:\n",
      " |      \n",
      " |      >>> print(rec[-10:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 0\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('IIEQFIDYLR', IUPACProtein())\n",
      " |      \n",
      " |      If you omit both, then you get a copy of the original record (although\n",
      " |      lacking the annotations and dbxrefs):\n",
      " |      \n",
      " |      >>> print(rec[:])\n",
      " |      ID: 1JOY\n",
      " |      Name: EnvZ\n",
      " |      Description: Homodimeric domain of EnvZ from E. coli\n",
      " |      Number of features: 1\n",
      " |      Per letter annotation for: secondary_structure\n",
      " |      Seq('MAAGVKQLADDRTLLMAGVSHDLRTPLTRIRLATEMMSEQDGYLAESINKDIEE...YLR', IUPACProtein())\n",
      " |      \n",
      " |      Finally, indexing with a simple integer is shorthand for pulling out\n",
      " |      that letter from the sequence directly:\n",
      " |      \n",
      " |      >>> rec[5]\n",
      " |      'K'\n",
      " |      >>> rec.seq[5]\n",
      " |      'K'\n",
      " |  \n",
      " |  __init__(self, seq, id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=None, features=None, annotations=None, letter_annotations=None)\n",
      " |      Create a SeqRecord.\n",
      " |      \n",
      " |      Arguments:\n",
      " |          - seq         - Sequence, required (Seq, MutableSeq or UnknownSeq)\n",
      " |          - id          - Sequence identifier, recommended (string)\n",
      " |          - name        - Sequence name, optional (string)\n",
      " |          - description - Sequence description, optional (string)\n",
      " |          - dbxrefs     - Database cross references, optional (list of strings)\n",
      " |          - features    - Any (sub)features, optional (list of SeqFeature objects)\n",
      " |          - annotations - Dictionary of annotations for the whole sequence\n",
      " |          - letter_annotations - Dictionary of per-letter-annotations, values\n",
      " |            should be strings, list or tuples of the same\n",
      " |            length as the full sequence.\n",
      " |      \n",
      " |      You will typically use Bio.SeqIO to read in sequences from files as\n",
      " |      SeqRecord objects.  However, you may want to create your own SeqRecord\n",
      " |      objects directly.\n",
      " |      \n",
      " |      Note that while an id is optional, we strongly recommend you supply a\n",
      " |      unique id string for each record.  This is especially important\n",
      " |      if you wish to write your sequences to a file.\n",
      " |      \n",
      " |      If you don't have the actual sequence, but you do know its length,\n",
      " |      then using the UnknownSeq object from Bio.Seq is appropriate.\n",
      " |      \n",
      " |      You can create a 'blank' SeqRecord object, and then populate the\n",
      " |      attributes later.\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Iterate over the letters in the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a protein FASTA file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/loveliesbleeding.pro\", \"fasta\")\n",
      " |      >>> for amino in record:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      This is just a shortcut for iterating over the sequence directly:\n",
      " |      \n",
      " |      >>> for amino in record.seq:\n",
      " |      ...     print(amino)\n",
      " |      ...     if amino == \"L\": break\n",
      " |      X\n",
      " |      A\n",
      " |      G\n",
      " |      L\n",
      " |      >>> print(record.seq[3])\n",
      " |      L\n",
      " |      \n",
      " |      Note that this does not facilitate iteration together with any\n",
      " |      per-letter-annotation.  However, you can achieve that using the\n",
      " |      python zip function on the record (or its sequence) and the relevant\n",
      " |      per-letter-annotation:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> rec = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(rec.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> for nuc, qual in zip(rec, rec.letter_annotations[\"solexa_quality\"]):\n",
      " |      ...     if qual > 35:\n",
      " |      ...         print(\"%s %i\" % (nuc, qual))\n",
      " |      A 40\n",
      " |      C 39\n",
      " |      G 38\n",
      " |      T 37\n",
      " |      A 36\n",
      " |      \n",
      " |      You may agree that using zip(rec.seq, ...) is more explicit than using\n",
      " |      zip(rec, ...) as shown above.\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Returns the length of the sequence.\n",
      " |      \n",
      " |      For example, using Bio.SeqIO to read in a FASTA nucleotide file:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/sweetpea.nu\", \"fasta\")\n",
      " |      >>> len(record)\n",
      " |      309\n",
      " |      >>> len(record.seq)\n",
      " |      309\n",
      " |  \n",
      " |  __nonzero__ = __bool__(self)\n",
      " |  \n",
      " |  __radd__(self, other)\n",
      " |      Add another sequence or string to this sequence (from the left).\n",
      " |      \n",
      " |      This method handles adding a Seq object (or similar, e.g. MutableSeq)\n",
      " |      or a plain Python string (on the left) to a SeqRecord (on the right).\n",
      " |      See the __add__ method for more details, but for example:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      \n",
      " |      >>> new = \"ACT\" + record\n",
      " |      >>> print(\"%s %s\" % (new.id, new.seq))\n",
      " |      slxa_0001_1_0001_01 ACTACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(new.letter_annotations))\n",
      " |      []\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      A concise summary of the record for debugging (string).\n",
      " |      \n",
      " |      The python built in function repr works by calling the object's ___repr__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Alphabet import generic_protein\n",
      " |      >>> rec = SeqRecord(Seq(\"MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKAT\"\n",
      " |      ...                    +\"GEMKEQTEWHRVVLFGKLAEVASEYLRKGSQVYIEGQLRTRKWTDQ\"\n",
      " |      ...                    +\"SGQDRYTTEVVVNVGGTMQMLGGRQGGGAPAGGNIGGGQPQGGWGQ\"\n",
      " |      ...                    +\"PQQPQGGNQFSGGAQSRPQQSAPAAPSNEPPMDFDDDIPF\",\n",
      " |      ...                    generic_protein),\n",
      " |      ...                 id=\"NP_418483.1\", name=\"b4059\",\n",
      " |      ...                 description=\"ssDNA-binding protein\",\n",
      " |      ...                 dbxrefs=[\"ASAP:13298\", \"GI:16131885\", \"GeneID:948570\"])\n",
      " |      >>> print(repr(rec))\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF', ProteinAlphabet()), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      At the python prompt you can also use this shorthand:\n",
      " |      \n",
      " |      >>> rec\n",
      " |      SeqRecord(seq=Seq('MASRGVNKVILVGNLGQDPEVRYMPNGGAVANITLATSESWRDKATGEMKEQTE...IPF', ProteinAlphabet()), id='NP_418483.1', name='b4059', description='ssDNA-binding protein', dbxrefs=['ASAP:13298', 'GI:16131885', 'GeneID:948570'])\n",
      " |      \n",
      " |      Note that long sequences are shown truncated. Also note that any\n",
      " |      annotations, letter_annotations and features are not shown (as they\n",
      " |      would lead to a very long string).\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      A human readable summary of the record and its annotation (string).\n",
      " |      \n",
      " |      The python built in function str works by calling the object's ___str__\n",
      " |      method.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      " |      ...                         IUPAC.protein),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein, small\")\n",
      " |      >>> print(str(record))\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      " |      \n",
      " |      In this example you don't actually need to call str explicity, as the\n",
      " |      print command does this automatically:\n",
      " |      \n",
      " |      >>> print(record)\n",
      " |      ID: YP_025292.1\n",
      " |      Name: HokC\n",
      " |      Description: toxic membrane protein, small\n",
      " |      Number of features: 0\n",
      " |      Seq('MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF', IUPACProtein())\n",
      " |      \n",
      " |      Note that long sequences are shown truncated.\n",
      " |  \n",
      " |  format(self, format)\n",
      " |      Returns the record as a string in the specified file format.\n",
      " |      \n",
      " |      The format should be a lower case string supported as an output\n",
      " |      format by Bio.SeqIO, which is used to turn the SeqRecord into a\n",
      " |      string.  e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> record = SeqRecord(Seq(\"MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\",\n",
      " |      ...                         IUPAC.protein),\n",
      " |      ...                    id=\"YP_025292.1\", name=\"HokC\",\n",
      " |      ...                    description=\"toxic membrane protein\")\n",
      " |      >>> record.format(\"fasta\")\n",
      " |      '>YP_025292.1 toxic membrane protein\\nMKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\\n'\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >YP_025292.1 toxic membrane protein\n",
      " |      MKQHKAMIVALIVICITAVVAALVTRKDLCEVHIRTGQTEVAVF\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      The python print command automatically appends a new line, meaning\n",
      " |      in this example a blank line is shown.  If you look at the string\n",
      " |      representation you can see there is a trailing new line (shown as\n",
      " |      slash n) which is important when writing to a file or if\n",
      " |      concatenating multiple sequence strings together.\n",
      " |      \n",
      " |      Note that this method will NOT work on every possible file format\n",
      " |      supported by Bio.SeqIO (e.g. some are for multiple sequences only).\n",
      " |  \n",
      " |  lower(self)\n",
      " |      Returns a copy of the record with a lower case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Fasta/aster.pro\", \"fasta\")\n",
      " |      >>> print(record.format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      GGHVNPAVTFGAFVGGNITLLRGIVYIIAQLLGSTVACLLLKFVTNDMAVGVFSLSAGVG\n",
      " |      VTNALVFEIVMTFGLVYTVYATAIDPKKGSLGTIAPIAIGFIVGANI\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(record.lower().format(\"fasta\"))\n",
      " |      >gi|3298468|dbj|BAA31520.1| SAMIPF\n",
      " |      gghvnpavtfgafvggnitllrgivyiiaqllgstvaclllkfvtndmavgvfslsagvg\n",
      " |      vtnalvfeivmtfglvytvyataidpkkgslgtiapiaigfivgani\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      To take a more annotation rich example,\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> old = SeqIO.read(\"EMBL/TRBG361.embl\", \"embl\")\n",
      " |      >>> len(old.features)\n",
      " |      3\n",
      " |      >>> new = old.lower()\n",
      " |      >>> len(old.features) == len(new.features)\n",
      " |      True\n",
      " |      >>> old.annotations[\"organism\"] == new.annotations[\"organism\"]\n",
      " |      True\n",
      " |      >>> old.dbxrefs == new.dbxrefs\n",
      " |      True\n",
      " |  \n",
      " |  reverse_complement(self, id=False, name=False, description=False, features=True, annotations=False, letter_annotations=True, dbxrefs=False)\n",
      " |      Returns new SeqRecord with reverse complement sequence.\n",
      " |      \n",
      " |      You can specify the returned record's id, name and description as\n",
      " |      strings, or True to keep that of the parent, or False for a default.\n",
      " |      \n",
      " |      You can specify the returned record's features with a list of\n",
      " |      SeqFeature objects, or True to keep that of the parent, or False to\n",
      " |      omit them. The default is to keep the original features (with the\n",
      " |      strand and locations adjusted).\n",
      " |      \n",
      " |      You can also specify both the returned record's annotations and\n",
      " |      letter_annotations as dictionaries, True to keep that of the parent,\n",
      " |      or False to omit them. The default is to keep the original\n",
      " |      annotations (with the letter annotations reversed).\n",
      " |      \n",
      " |      To show what happens to the pre-letter annotations, consider an\n",
      " |      example Solexa variant FASTQ file with a single entry, which we'll\n",
      " |      read in as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Now take the reverse complement,\n",
      " |      \n",
      " |      >>> rc_record = record.reverse_complement(id=record.id+\"_rc\")\n",
      " |      >>> print(\"%s %s\" % (rc_record.id, rc_record.seq))\n",
      " |      slxa_0001_1_0001_01_rc NNNNNNACGTACGTACGTACGTACGTACGTACGTACGTACGTACGT\n",
      " |      \n",
      " |      Notice that the per-letter-annotations have also been reversed,\n",
      " |      although this may not be appropriate for all cases.\n",
      " |      \n",
      " |      >>> print(rc_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40]\n",
      " |      \n",
      " |      Now for the features, we need a different example. Parsing a GenBank\n",
      " |      file is probably the easiest way to get an nice example with features\n",
      " |      in it...\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> with open(\"GenBank/pBAD30.gb\") as handle:\n",
      " |      ...     plasmid = SeqIO.read(handle, \"gb\")\n",
      " |      >>> print(\"%s %i\" % (plasmid.id, len(plasmid)))\n",
      " |      pBAD30 4923\n",
      " |      >>> plasmid.seq\n",
      " |      Seq('GCTAGCGGAGTGTATACTGGCTTACTATGTTGGCACTGATGAGGGTGTCAGTGA...ATG', IUPACAmbiguousDNA())\n",
      " |      >>> len(plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Now, let's take the reverse complement of this whole plasmid:\n",
      " |      \n",
      " |      >>> rc_plasmid = plasmid.reverse_complement(id=plasmid.id+\"_rc\")\n",
      " |      >>> print(\"%s %i\" % (rc_plasmid.id, len(rc_plasmid)))\n",
      " |      pBAD30_rc 4923\n",
      " |      >>> rc_plasmid.seq\n",
      " |      Seq('CATGGGCAAATATTATACGCAAGGCGACAAGGTGCTGATGCCGCTGGCGATTCA...AGC', IUPACAmbiguousDNA())\n",
      " |      >>> len(rc_plasmid.features)\n",
      " |      13\n",
      " |      \n",
      " |      Let's compare the first CDS feature - it has gone from being the\n",
      " |      second feature (index 1) to the second last feature (index -2), its\n",
      " |      strand has changed, and the location switched round.\n",
      " |      \n",
      " |      >>> print(plasmid.features[1])\n",
      " |      type: CDS\n",
      " |      location: [1081:1960](-)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(rc_plasmid.features[-2])\n",
      " |      type: CDS\n",
      " |      location: [2963:3842](+)\n",
      " |      qualifiers:\n",
      " |          Key: label, Value: ['araC']\n",
      " |          Key: note, Value: ['araC regulator of the arabinose BAD promoter']\n",
      " |          Key: vntifkey, Value: ['4']\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      You can check this new location, based on the length of the plasmid:\n",
      " |      \n",
      " |      >>> len(plasmid) - 1081\n",
      " |      3842\n",
      " |      >>> len(plasmid) - 1960\n",
      " |      2963\n",
      " |      \n",
      " |      Note that if the SeqFeature annotation includes any strand specific\n",
      " |      information (e.g. base changes for a SNP), this information is not\n",
      " |      amended, and would need correction after the reverse complement.\n",
      " |      \n",
      " |      Note trying to reverse complement a protein SeqRecord raises an\n",
      " |      exception:\n",
      " |      \n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.Alphabet import IUPAC\n",
      " |      >>> protein_rec = SeqRecord(Seq(\"MAIVMGR\", IUPAC.protein), id=\"Test\")\n",
      " |      >>> protein_rec.reverse_complement()\n",
      " |      Traceback (most recent call last):\n",
      " |         ...\n",
      " |      ValueError: Proteins do not have complements!\n",
      " |      \n",
      " |      Also note you can reverse complement a SeqRecord using a MutableSeq:\n",
      " |      \n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Seq import MutableSeq\n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> rec = SeqRecord(MutableSeq(\"ACGT\", generic_dna), id=\"Test\")\n",
      " |      >>> rec.seq[0] = \"T\"\n",
      " |      >>> print(\"%s %s\" % (rec.id, rec.seq))\n",
      " |      Test TCGT\n",
      " |      >>> rc = rec.reverse_complement(id=True)\n",
      " |      >>> print(\"%s %s\" % (rc.id, rc.seq))\n",
      " |      Test ACGA\n",
      " |  \n",
      " |  upper(self)\n",
      " |      Returns a copy of the record with an upper case sequence.\n",
      " |      \n",
      " |      All the annotation is preserved unchanged. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Alphabet import generic_dna\n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> record = SeqRecord(Seq(\"acgtACGT\", generic_dna), id=\"Test\",\n",
      " |      ...                    description = \"Made up for this example\")\n",
      " |      >>> record.letter_annotations[\"phred_quality\"] = [1, 2, 3, 4, 5, 6, 7, 8]\n",
      " |      >>> print(record.upper().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      ACGTACGT\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      Naturally, there is a matching lower method:\n",
      " |      \n",
      " |      >>> print(record.lower().format(\"fastq\"))\n",
      " |      @Test Made up for this example\n",
      " |      acgtacgt\n",
      " |      +\n",
      " |      \"#$%&'()\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  letter_annotations\n",
      " |      Dictionary of per-letter-annotation for the sequence.\n",
      " |      \n",
      " |      For example, this can hold quality scores used in FASTQ or QUAL files.\n",
      " |      Consider this example using Bio.SeqIO to read in an example Solexa\n",
      " |      variant FASTQ file as a SeqRecord:\n",
      " |      \n",
      " |      >>> from Bio import SeqIO\n",
      " |      >>> record = SeqIO.read(\"Quality/solexa_faked.fastq\", \"fastq-solexa\")\n",
      " |      >>> print(\"%s %s\" % (record.id, record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTACGTACGTACGTACGTACGTACGTACGTACGTACGTNNNNNN\n",
      " |      >>> print(list(record.letter_annotations))\n",
      " |      ['solexa_quality']\n",
      " |      >>> print(record.letter_annotations[\"solexa_quality\"])\n",
      " |      [40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      The letter_annotations get sliced automatically if you slice the\n",
      " |      parent SeqRecord, for example taking the last ten bases:\n",
      " |      \n",
      " |      >>> sub_record = record[-10:]\n",
      " |      >>> print(\"%s %s\" % (sub_record.id, sub_record.seq))\n",
      " |      slxa_0001_1_0001_01 ACGTNNNNNN\n",
      " |      >>> print(sub_record.letter_annotations[\"solexa_quality\"])\n",
      " |      [4, 3, 2, 1, 0, -1, -2, -3, -4, -5]\n",
      " |      \n",
      " |      Any python sequence (i.e. list, tuple or string) can be recorded in\n",
      " |      the SeqRecord's letter_annotations dictionary as long as the length\n",
      " |      matches that of the SeqRecord's sequence.  e.g.\n",
      " |      \n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      1\n",
      " |      >>> sub_record.letter_annotations[\"dummy\"] = \"abcdefghij\"\n",
      " |      >>> len(sub_record.letter_annotations)\n",
      " |      2\n",
      " |      \n",
      " |      You can delete entries from the letter_annotations dictionary as usual:\n",
      " |      \n",
      " |      >>> del sub_record.letter_annotations[\"solexa_quality\"]\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {'dummy': 'abcdefghij'}\n",
      " |      \n",
      " |      You can completely clear the dictionary easily as follows:\n",
      " |      \n",
      " |      >>> sub_record.letter_annotations = {}\n",
      " |      >>> sub_record.letter_annotations\n",
      " |      {}\n",
      " |  \n",
      " |  seq\n",
      " |      The sequence itself, as a Seq or MutableSeq object.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dir(tp53_rec)\n",
    "help(tp53_rec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dbxrefs\n",
      "description\n",
      "annotations\n",
      "features\n",
      "id\n",
      "name\n"
     ]
    }
   ],
   "source": [
    "#print tp53_rec.__dict__\n",
    "print tp53_rec.description\n",
    "\n",
    "for key in tp53_rec.__dict__:\n",
    "    if not key.startswith('_'):\n",
    "        print key, str(tp53_rec.__getattribute__(key))[:50]+\"...\"\n",
    "#for key,val in tp53_rec.annotations.iteritems():\n",
    "#    print key, \" \"*15, str(val)[:50]+\"...\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: the information you get depends on the format. Let's try with fasta to see that we get next to nothing, compared to the genbank format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">gi|120407068|ref|NP_000537.3| cellular tumor antigen p53 isoform a [Homo sapiens]\n",
      "MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAA\n",
      "PPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKT\n",
      "CPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRN\n",
      "TFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGR\n",
      "DRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALEL\n",
      "KDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD\n",
      "\n",
      "\n",
      "dbxrefs []...\n",
      "description gi|120407068|ref|NP_000537.3| cellular tumor antig...\n",
      "annotations {}...\n",
      "features []...\n",
      "id gi|120407068|ref|NP_000537.3|...\n",
      "name gi|120407068|ref|NP_000537.3|...\n"
     ]
    }
   ],
   "source": [
    "print Entrez.efetch(db=\"protein\",id=[\"NP_000537\"],rettype=\"fasta\").read()\n",
    "tp53_rec_fasta=SeqIO.read(Entrez.efetch(db=\"protein\",id=[\"NP_000537\"],rettype=\"fasta\"),\"fasta\")\n",
    "for key in tp53_rec_fasta.__dict__:\n",
    "    if not key.startswith('_'):\n",
    "        print key, str(tp53_rec_fasta.__getattribute__(key))[:50]+\"...\"\n",
    "for key,val in tp53_rec_fasta.annotations.iteritems():\n",
    "    print key, \" \"*15, str(val)[:50]+\"...\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Features\n",
    "\n",
    "Annotated features along the sequence. Best seen in the [source](http://www.ncbi.nlm.nih.gov/nuccore/NM_001276697.1). The complete list of possible feature types is [here](http://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html#FeaturesB)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "source [0:393] {'db_xref': ['taxon:9606'], 'organism': ['Homo sapiens'], 'chromosome': ['17'], 'map': ['17p13.1']}\n",
      "Protein [0:393] {'note': ['tumor protein 53; mutant tumor protein 53; cellular tumor antigen p53; phosphoprotein p53; transformation-related protein 53; p53 tumor suppressor; antigen NY-CO-13; tumor supressor p53'], 'product': ['cellular tumor antigen p53 isoform a'], 'calculated_mol_wt': ['43522']}\n",
      "Region [0:320] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interaction with CCAR2. {ECO:0000269|PubMed:25732823}']}\n",
      "Region [0:83] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interaction with HRMT1L2']}\n",
      "Region [0:44] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Transcription activation (acidic)']}\n",
      "Region [4:29] {'note': ['P53 transactivation motif; pfam08563'], 'db_xref': ['CDD:149567'], 'region_name': ['P53_TAD']}\n",
      "Site [5:6] {'db_xref': ['HPRD:02739'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[50]'], 'site_type': ['phosphorylation']}\n",
      "Site [8:9] {'note': ['Phosphoserine, by HIPK4. {ECO:0000269|PubMed:18022393}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [8:9] {'db_xref': ['HPRD:02739'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[27]', '[50]'], 'site_type': ['phosphorylation']}\n",
      "Site [8:9] {'db_xref': ['HPRD:06347'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[27]', '[50]'], 'site_type': ['phosphorylation']}\n",
      "Site [14:15] {'note': ['dephosphorylation site'], 'db_xref': ['HPRD:01146'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[11]'], 'site_type': ['modified']}\n",
      "Site [14:15] {'note': ['Phosphoserine, by CDK5, PRPK, AMPK, NUAK1 and ATM. {ECO:0000269|PubMed:10570149, ECO:0000269|PubMed:11554766, ECO:0000269|PubMed:15701641, ECO:0000269|PubMed:15866171, ECO:0000269|PubMed:17108107, ECO:0000269|PubMed:17591690, ECO:0000269|PubMed:17967874, ECO:0000269|PubMed:21317932}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [14:15] {'db_xref': ['HPRD:01496'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[19]', '[25]'], 'site_type': ['phosphorylation']}\n",
      "Site [14:15] {'db_xref': ['HPRD:02941'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[31]', '[33]', '[44]', '[50]'], 'site_type': ['phosphorylation']}\n",
      "Site [14:15] {'db_xref': ['HPRD:04356'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[31]', '[33]', '[44]', '[50]'], 'site_type': ['phosphorylation']}\n",
      "Site [14:15] {'db_xref': ['HPRD:06347'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[10]', '[17]', '[27]', '[31]', '[33]', '[43]', '[44]', '[49]', '[50]'], 'site_type': ['phosphorylation']}\n",
      "Site [14:15] {'db_xref': ['HPRD:08369'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[14]', '[17]', '[31]', '[33]', '[44]', '[50]'], 'site_type': ['phosphorylation']}\n",
      "Site [14:15] {'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[16]', '[43]'], 'site_type': ['phosphorylation']}\n",
      "Region [16:25] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['TADI']}\n",
      "Site [17:18] {'note': ['proteolytic cleavage site'], 'db_xref': ['HPRD:00253'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[26]'], 'site_type': ['modified']}\n",
      "Site [17:18] {'note': ['Phosphothreonine, by CK1, VRK1 and VRK2. {ECO:0000269|PubMed:10606744, ECO:0000269|PubMed:10951572, ECO:0000269|PubMed:16704422}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [17:18] {'db_xref': ['HPRD:02920'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[46]'], 'site_type': ['phosphorylation']}\n",
      "Site [17:18] {'db_xref': ['HPRD:02941'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[29]', '[33]', '[39]', '[42]'], 'site_type': ['phosphorylation']}\n",
      "Site [17:18] {'db_xref': ['HPRD:03701'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[29]', '[33]', '[39]', '[42]'], 'site_type': ['phosphorylation']}\n",
      "Site [17:18] {'db_xref': ['HPRD:08369'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[29]', '[31]', '[33]', '[39]', '[42]', '[44]'], 'site_type': ['phosphorylation']}\n",
      "Site [17:18] {'db_xref': ['HPRD:00277'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[29]', '[33]', '[39]', '[42]'], 'site_type': ['phosphorylation']}\n",
      "Site [17:18] {'db_xref': ['HPRD:00278'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[29]', '[33]', '[39]', '[42]'], 'site_type': ['phosphorylation']}\n",
      "Site [19:20] {'note': ['proteolytic cleavage site'], 'db_xref': ['HPRD:00253'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[26]'], 'site_type': ['modified']}\n",
      "Site [19:20] {'note': ['Phosphoserine, by CHEK2, CK1 and PLK3. {ECO:0000269|PubMed:10570149, ECO:0000269|PubMed:11447225, ECO:0000269|PubMed:11551930, ECO:0000269|PubMed:12810724, ECO:0000269|PubMed:20041275}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [19:20] {'db_xref': ['HPRD:02941'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[31]', '[33]', '[44]'], 'site_type': ['phosphorylation']}\n",
      "Site [19:20] {'db_xref': ['HPRD:04222'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[22]', '[24]', '[31]', '[33]', '[44]'], 'site_type': ['phosphorylation']}\n",
      "Site [19:20] {'db_xref': ['HPRD:04356'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[31]', '[33]', '[44]'], 'site_type': ['phosphorylation']}\n",
      "Site [19:20] {'db_xref': ['HPRD:06347'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[27]', '[31]', '[33]', '[44]'], 'site_type': ['phosphorylation']}\n",
      "Site [19:20] {'db_xref': ['HPRD:08369'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[31]', '[33]', '[44]'], 'site_type': ['phosphorylation']}\n",
      "Site [22:23] {'note': ['proteolytic cleavage site'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[53]'], 'site_type': ['modified']}\n",
      "Site [32:33] {'note': ['Phosphoserine, by CDK5 and CDK7. {ECO:0000269|PubMed:17591690, ECO:0000269|PubMed:9372954}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [32:33] {'db_xref': ['HPRD:02941'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[33]'], 'site_type': ['phosphorylation']}\n",
      "Site [32:33] {'db_xref': ['HPRD:05418'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[28]', '[36]'], 'site_type': ['phosphorylation']}\n",
      "Site [32:33] {'db_xref': ['HPRD:06347'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[33]'], 'site_type': ['phosphorylation']}\n",
      "Site [32:33] {'db_xref': ['HPRD:00449'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[28]', '[36]'], 'site_type': ['phosphorylation']}\n",
      "Site [32:33] {'db_xref': ['HPRD:16016'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[12]'], 'site_type': ['phosphorylation']}\n",
      "Site [32:33] {'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[23]'], 'site_type': ['phosphorylation']}\n",
      "Site [36:37] {'note': ['Phosphoserine, by MAPKAPK5. {ECO:0000269|PubMed:17254968}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [36:37] {'db_xref': ['HPRD:02941'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[33]', '[44]'], 'site_type': ['phosphorylation']}\n",
      "Site [36:37] {'db_xref': ['HPRD:04356'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[33]', '[44]'], 'site_type': ['phosphorylation']}\n",
      "Site [36:37] {'db_xref': ['HPRD:06347'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[33]', '[44]'], 'site_type': ['phosphorylation']}\n",
      "Site [36:37] {'db_xref': ['HPRD:08369'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[33]', '[44]'], 'site_type': ['phosphorylation']}\n",
      "Site [36:37] {'db_xref': ['HPRD:09467'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[5]'], 'site_type': ['phosphorylation']}\n",
      "Region [39:40] {'note': ['delta40p53alpha'], 'region_name': ['alternative start codon']}\n",
      "Site [45:46] {'note': ['Phosphoserine, by CDK5, DYRK2, HIPK2 and PKC/PRKCG. {ECO:0000269|PubMed:11740489, ECO:0000269|PubMed:11780126, ECO:0000269|PubMed:16377624, ECO:0000269|PubMed:17349958, ECO:0000269|PubMed:17591690}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [45:46] {'db_xref': ['HPRD:06039'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[15]', '[27]', '[30]'], 'site_type': ['phosphorylation']}\n",
      "Site [45:46] {'db_xref': ['HPRD:06347'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[27]', '[30]'], 'site_type': ['phosphorylation']}\n",
      "Region [47:56] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['TADII']}\n",
      "Site [54:55] {'note': ['dephosphorylation site'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[6]'], 'site_type': ['modified']}\n",
      "Site [54:55] {'note': ['Phosphothreonine, by TAF1 and GRK5. {ECO:0000269|PubMed:15053879, ECO:0000269|PubMed:20124405}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [54:55] {'db_xref': ['HPRD:01496'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[18]', '[34]'], 'site_type': ['phosphorylation']}\n",
      "Region [65:110] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interaction with WWOX']}\n",
      "Site [80:81] {'db_xref': ['HPRD:03100'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[35]'], 'site_type': ['phosphorylation']}\n",
      "Site [80:81] {'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[23]'], 'site_type': ['phosphorylation']}\n",
      "Site [98:99] {'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[4]'], 'site_type': ['phosphorylation']}\n",
      "Region [99:370] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interaction with HIPK1. {ECO:0000250}']}\n",
      "Region [99:300] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Required for interaction with ZNF385A']}\n",
      "Region [108:288] {'note': ['P53 DNA-binding domain; cd08367'], 'db_xref': ['CDD:176262'], 'region_name': ['P53']}\n",
      "Region [112:236] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Required for interaction with FBXO42']}\n",
      "Region [115:292] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interaction with AXIN1. {ECO:0000250}']}\n",
      "Site [119:120] {'note': ['N6-acetyllysine, by KAT6A. {ECO:0000269|PubMed:23431171}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['acetylation']}\n",
      "Site [119:120] {'note': ['Interaction with DNA; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['other']}\n",
      "Site [154:155] {'db_xref': ['HPRD:00277'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[21]'], 'site_type': ['phosphorylation']}\n",
      "Site order{[175:176], [178:179], [237:238], [241:242]} {'note': ['zinc binding site [ion binding]'], 'db_xref': ['CDD:176262'], 'site_type': ['other']}\n",
      "Site order{[176:179], [180:181]} {'note': ['dimerization site [polypeptide binding]'], 'db_xref': ['CDD:176262'], 'site_type': ['other']}\n",
      "Site [182:183] {'note': ['Phosphoserine, by AURKB. {ECO:0000269|PubMed:20959462}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site order{[238:239], [240:241], [247:248], [272:273], [274:277], [279:280]} {'note': ['DNA binding site [nucleotide binding]'], 'db_xref': ['CDD:176262'], 'site_type': ['DNA binding']}\n",
      "Region [240:248] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interacts with the 53BP2 SH3 domain']}\n",
      "Region [255:294] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interaction with E4F1']}\n",
      "Site [268:269] {'note': ['Phosphoserine, by AURKB. {ECO:0000269|PubMed:20959462}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Region [272:280] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interaction with DNA']}\n",
      "Site [283:284] {'note': ['Phosphothreonine, by AURKB. {ECO:0000269|PubMed:20959462}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [291:292] {'db_xref': ['HPRD:04078'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[20]'], 'site_type': ['acetylation']}\n",
      "Region [299:393] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interaction with CARM1']}\n",
      "Region [304:321] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Bipartite nuclear localization signal']}\n",
      "Site [304:305] {'note': ['N6-acetyllysine. {ECO:0000269|PubMed:12724314}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['acetylation']}\n",
      "Site [304:305] {'db_xref': ['HPRD:04078'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[20]'], 'site_type': ['acetylation']}\n",
      "Site [314:315] {'note': ['dephosphorylation site'], 'db_xref': ['HPRD:04614'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[45]'], 'site_type': ['modified']}\n",
      "Site [314:315] {'note': ['dephosphorylation site'], 'db_xref': ['HPRD:04615'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[45]'], 'site_type': ['modified']}\n",
      "Site [314:315] {'note': ['Phosphoserine, by AURKA, CDK1 and CDK2. {ECO:0000269|PubMed:10884347, ECO:0000269|PubMed:14702041}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [314:315] {'db_xref': ['HPRD:00310'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[28]', '[40]'], 'site_type': ['phosphorylation']}\n",
      "Site [314:315] {'db_xref': ['HPRD:00449'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[28]', '[40]'], 'site_type': ['phosphorylation']}\n",
      "Site [314:315] {'db_xref': ['HPRD:16016'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[12]'], 'site_type': ['phosphorylation']}\n",
      "Site [314:315] {'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[23]'], 'site_type': ['phosphorylation']}\n",
      "Region [318:360] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interaction with HIPK2']}\n",
      "Site [318:319] {'note': ['ubiquitination site'], 'db_xref': ['HPRD:04317'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[9]'], 'site_type': ['modified']}\n",
      "Site [319:320] {'db_xref': ['HPRD:06780'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[9]', '[48]', '[51]'], 'site_type': ['acetylation']}\n",
      "Site [319:320] {'note': ['ubiquitination site'], 'db_xref': ['HPRD:04317'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[9]'], 'site_type': ['modified']}\n",
      "Site [319:320] {'note': ['neddylation site'], 'db_xref': ['HPRD:06985'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[7]'], 'site_type': ['modified']}\n",
      "Site [320:321] {'note': ['N6-acetyllysine. {ECO:0000250|UniProtKB:P02340}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['acetylation']}\n",
      "Site [320:321] {'note': ['ubiquitination site'], 'db_xref': ['HPRD:04317'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[9]'], 'site_type': ['modified']}\n",
      "Site [320:321] {'note': ['neddylation site'], 'db_xref': ['HPRD:06985'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[7]'], 'site_type': ['modified']}\n",
      "Region [323:359] {'note': ['P53 tetramerization motif; pfam07710'], 'db_xref': ['CDD:254376'], 'region_name': ['P53_tetramer']}\n",
      "Region [324:356] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Oligomerization']}\n",
      "Region [338:350] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Nuclear export signal']}\n",
      "Site [350:351] {'note': ['ubiquitination site'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[38]'], 'site_type': ['modified']}\n",
      "Site [356:357] {'note': ['ubiquitination site'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[38]'], 'site_type': ['modified']}\n",
      "Region [358:363] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Interaction with USP7']}\n",
      "Site [365:366] {'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[55]'], 'site_type': ['phosphorylation']}\n",
      "Region [367:387] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['Basic (repression of DNA-binding)']}\n",
      "Region [369:372] {'note': ['propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'region_name': ['[KR]-[STA]-K motif']}\n",
      "Site [369:370] {'db_xref': ['HPRD:04078'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[54]'], 'site_type': ['acetylation']}\n",
      "Site [369:370] {'note': ['N6,N6-dimethyllysine, alternate. {ECO:0000269|PubMed:17108971, ECO:0000269|PubMed:22864287}; N6-methyllysine, by SMYD2, alternate. {ECO:0000269|PubMed:17108971, ECO:0000269|PubMed:22864287}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['methylation']}\n",
      "Site [369:370] {'db_xref': ['HPRD:15406'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[8]'], 'site_type': ['methylation']}\n",
      "Site [369:370] {'note': ['ubiquitination site'], 'db_xref': ['HPRD:01272'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[9]'], 'site_type': ['modified']}\n",
      "Site [369:370] {'note': ['neddylation site'], 'db_xref': ['HPRD:06985'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[7]'], 'site_type': ['modified']}\n",
      "Site [370:371] {'db_xref': ['HPRD:01498'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[52]'], 'site_type': ['phosphorylation']}\n",
      "Site [371:372] {'db_xref': ['HPRD:04078'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[54]'], 'site_type': ['acetylation']}\n",
      "Site [371:372] {'note': ['N6-methyllysine, by SETD7. {ECO:0000269|PubMed:15525938, ECO:0000269|PubMed:16415881}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['methylation']}\n",
      "Site [371:372] {'db_xref': ['HPRD:16226'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[8]', '[13]'], 'site_type': ['methylation']}\n",
      "Site [371:372] {'note': ['ubiquitination site'], 'db_xref': ['HPRD:01272'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[9]'], 'site_type': ['modified']}\n",
      "Site [371:372] {'note': ['neddylation site'], 'db_xref': ['HPRD:06985'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[7]'], 'site_type': ['modified']}\n",
      "Site [372:373] {'note': ['N6-acetyllysine, alternate. {ECO:0000269|PubMed:10656795}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['acetylation']}\n",
      "Site [372:373] {'db_xref': ['HPRD:04078'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[54]'], 'site_type': ['acetylation']}\n",
      "Site [372:373] {'note': ['N6,N6-dimethyllysine, by EHMT1 and EHMT2, alternate. {ECO:0000269|PubMed:20118233}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['methylation']}\n",
      "Site [372:373] {'note': ['ubiquitination site'], 'db_xref': ['HPRD:01272'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[9]', '[37]'], 'site_type': ['modified']}\n",
      "Site [372:373] {'note': ['deacetylation site'], 'db_xref': ['HPRD:03143'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[41]'], 'site_type': ['modified']}\n",
      "Site [372:373] {'note': ['neddylation site'], 'db_xref': ['HPRD:06985'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[7]'], 'site_type': ['modified']}\n",
      "Site [377:378] {'db_xref': ['HPRD:01498'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[58]'], 'site_type': ['phosphorylation']}\n",
      "Site [380:381] {'note': ['N6-acetyllysine. {ECO:0000244|PubMed:19608861}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['acetylation']}\n",
      "Site [380:381] {'db_xref': ['HPRD:04078'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[54]'], 'site_type': ['acetylation']}\n",
      "Site [380:381] {'note': ['ubiquitination site'], 'db_xref': ['HPRD:01272'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[9]'], 'site_type': ['modified']}\n",
      "Site [380:381] {'note': ['neddylation site'], 'db_xref': ['HPRD:06985'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[7]'], 'site_type': ['modified']}\n",
      "Site [381:382] {'note': ['N6-acetyllysine, by KAT6A, alternate. {ECO:0000244|PubMed:19608861, ECO:0000269|PubMed:10656795, ECO:0000269|PubMed:20228809, ECO:0000269|PubMed:23431171}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['acetylation']}\n",
      "Site [381:382] {'db_xref': ['HPRD:04078'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[51]'], 'site_type': ['acetylation']}\n",
      "Site [381:382] {'note': ['N6,N6-dimethyllysine, alternate. {ECO:0000269|PubMed:17707234, ECO:0000269|PubMed:20870725, ECO:0000269|PubMed:22864287}; N6-methyllysine, by SETD8, alternate. {ECO:0000269|PubMed:17707234, ECO:0000269|PubMed:20870725, ECO:0000269|PubMed:22864287}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['methylation']}\n",
      "Site [381:382] {'note': ['deacylation site'], 'db_xref': ['HPRD:08381'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[32]'], 'site_type': ['modified']}\n",
      "Site [381:382] {'note': ['deacetylation site'], 'db_xref': ['HPRD:03143'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[41]'], 'site_type': ['modified']}\n",
      "Site [381:382] {'note': ['neddylation site'], 'db_xref': ['HPRD:06985'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[7]'], 'site_type': ['modified']}\n",
      "Site [385:386] {'note': ['ubiquitination site'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[9]', '[38]'], 'site_type': ['modified']}\n",
      "Site [385:386] {'note': ['ubiquitination site'], 'db_xref': ['HPRD:01272'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[9]'], 'site_type': ['modified']}\n",
      "Site [385:386] {'note': ['neddylation site'], 'db_xref': ['HPRD:06985'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[7]'], 'site_type': ['modified']}\n",
      "Site [391:392] {'note': ['Phosphoserine, by CK2, CDK2 and NUAK1. {ECO:0000269|PubMed:10884347, ECO:0000269|PubMed:11239457, ECO:0000269|PubMed:17108107, ECO:0000269|PubMed:21317932, ECO:0000269|PubMed:22214662}; propagated from UniProtKB/Swiss-Prot (P04637.4)'], 'experiment': ['experimental evidence, no additional details recorded'], 'site_type': ['phosphorylation']}\n",
      "Site [391:392] {'db_xref': ['HPRD:01468'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[40]', '[42]', '[47]'], 'site_type': ['phosphorylation']}\n",
      "Site [391:392] {'db_xref': ['HPRD:02619'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[40]', '[42]', '[47]', '[64]'], 'site_type': ['phosphorylation']}\n",
      "Site [391:392] {'db_xref': ['HPRD:00277'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[40]', '[42]', '[47]'], 'site_type': ['phosphorylation']}\n",
      "Site [391:392] {'db_xref': ['HPRD:00310'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[40]', '[42]', '[47]'], 'site_type': ['phosphorylation']}\n",
      "Site [391:392] {'db_xref': ['HPRD:16016'], 'experiment': ['experimental evidence, no additional details recorded'], 'citation': ['[12]'], 'site_type': ['phosphorylation']}\n",
      "CDS [0:393] {'note': ['isoform a is encoded by transcript variant 1'], 'coded_by': ['NM_000546.5:203..1384'], 'gene': ['TP53'], 'gene_synonym': ['BCC7; LFS1; P53; TRP53'], 'db_xref': ['CCDS:CCDS11118.1', 'GeneID:7157', 'HGNC:HGNC:11998', 'MIM:191170']}\n"
     ]
    }
   ],
   "source": [
    "for f in tp53_rec.features:\n",
    "    print f.type, f.location, f.qualifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CDS at [359:1064](+)\n",
      "SeqRecord(seq=Seq('ATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCC...TGA', IUPACAmbiguousDNA()), id='NM_001276697.1', name='NM_001276697', description='Homo sapiens tumor protein p53 (TP53), transcript variant 5, mRNA.', dbxrefs=[])\n",
      "ATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA\n"
     ]
    }
   ],
   "source": [
    "#extract all CDS from a gene\n",
    "tp53_rec_gene=SeqIO.read(Entrez.efetch(db=\"nucleotide\",id=[\"NM_001276697.1\"],rettype=\"gb\"),\"gb\")\n",
    "for f in tp53_rec_gene.features:\n",
    "    if f.type==\"CDS\":\n",
    "        print \"CDS at\", f.location\n",
    "        extracted=f.extract(tp53_rec_gene)\n",
    "        print repr(extracted)\n",
    "        print extracted.seq\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# File I/O\n",
    "\n",
    "`SeqIO.read()` and `.parse()` also work with file name or an open file. And `SeqIO.write(sequences,filename,format)` can write into a file in a [number of formats](http://biopython.org/wiki/SeqIO#File_Formats)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "records written: 1\n",
      "SeqRecord(seq=Seq('TGAGGCCAGGAGATGGAGGCTGCAGTGAGCTGTGATCACACCACTGTGCTCCAG...GTG', IUPACAmbiguousDNA()), id='NM_001276697.1', name='NM_001276697', description='Homo sapiens tumor protein p53 (TP53), transcript variant 5, mRNA.', dbxrefs=[])\n",
      "False\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "print \"records written:\", SeqIO.write([tp53_rec_gene],\"tp53.gb\",\"gb\")\n",
    "tp53_rec_gene_fromfile=SeqIO.read(\"tp53.gb\",\"gb\")\n",
    "print repr(tp53_rec_gene_fromfile)\n",
    "print tp53_rec_gene_fromfile==tp53_rec_gene #Why false?\n",
    "print tp53_rec_gene_fromfile.seq==tp53_rec_gene.seq #Why true?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lecture code\n",
    "\n",
    "...here I gather little examples shown during the lecture, not necessarily related to the lecture topic. :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "s=\"asbifdshbijiugfsd\"\n",
    "print \"bif\" in s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95]\n"
     ]
    }
   ],
   "source": [
    "print list(range(0,100,5))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.5+"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
